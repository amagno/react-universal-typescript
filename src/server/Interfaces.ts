

export interface RouteParamsInterface<PropsInterface = {}> {
  pathname: string;
  context?: PropsInterface;
  params?: any;
}

export interface RouteRenderInterface {

}