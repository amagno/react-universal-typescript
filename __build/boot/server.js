"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const BootServer_1 = require("../server/BootServer");
BootServer_1.default(3000);
