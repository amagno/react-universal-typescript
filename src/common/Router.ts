import { Routes, resolve } from 'universal-router';
import { RouteParamsInterface } from './CommonInterfaces';
import * as React from 'react';

const UniversalRouter = require('universal-router');

export default class Router {
  private routes: Routes<RouteParamsInterface, React.ReactElement<{}>>;

  constructor(routes: Routes<RouteParamsInterface, React.ReactElement<{}>> = []) {
    this.routes = routes;
  }
  public add(path: string, component: React.ComponentClass | React.StatelessComponent): Router {
    this.routes.push({
      path,
      action: ({ pathname, params, context }) => {
        const props_chunck: RouteParamsInterface = {
          pathname,
          params: params || {},
          context: context || {}

        };
        return React.createElement(component, <React.Attributes>{...props_chunck});
      }
    });
    return this;
  }
  public async resolve<Props>(resolveParams: RouteParamsInterface<Props>) {
    try {
      const router = new UniversalRouter(this.routes);
      return await router.resolve(resolveParams);
    } catch (error) {
      throw new Error(error);
    }
  }
}