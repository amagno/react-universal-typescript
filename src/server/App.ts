import * as express from 'express';
import MakeDocument from './Lib/MakeDocument';
const handleRender = (req: express.Request, res: express.Response) => {
  res.send(MakeDocument('Hello World'));
};
const configureRender = (app: express.Application, path: string = '*'): express.Application => {
  return app.get(path, handleRender);
};
class App {
  private app: express.Application;
  constructor(app: express.Application, options?: object) {
    this.app = configureRender(app);
  }
  express() {
    return this.app;
  }
}
export default App;