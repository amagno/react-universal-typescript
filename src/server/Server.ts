import * as express from 'express';
import * as http from 'http';
import { ConsoleLogger } from '../utils/Logger';

class Server {
  private http: http.Server;

  constructor(app: express.Application, options?: object) {
    this.http = http.createServer(app);
  }
  listen(port: number = 3000, address?: string) {
    this.http.listen(port, address, () => {
      const _port = this.http.address().port;
      const _address = this.http.address().address;
      ConsoleLogger.info(`Server is running http://${_address}:${_port}/`);
    });
  }
}

export default Server;