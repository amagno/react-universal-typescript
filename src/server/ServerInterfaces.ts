import * as express from 'express';
import { RouteProps } from 'react-router';


export interface AppInterface {
  express(): express.Application;
}
export interface AppMiddlewareInterface {
  path?: string;
  use: express.Handler;
  dev?: boolean;
}
export interface BootOptionsInterface {
  routes: RouteProps[];
  port?: number;
  isDev?: boolean;
  app: express.Application;
  middlewares: AppMiddlewareInterface[];
}


