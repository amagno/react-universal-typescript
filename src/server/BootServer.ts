import Server from './Server';
import App from './App';
import * as express from 'express';
export default function (port: number): express.Application {
  const app = new App(express());
  const server = new Server(app.express());
  server.listen(port);
  return app.express();
}