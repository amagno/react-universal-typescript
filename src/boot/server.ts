import BootServer from '../server/BootServer';


BootServer(3000);

