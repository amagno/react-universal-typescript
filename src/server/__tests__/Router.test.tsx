import 'jest';
import * as React from 'react';
import { Routes } from 'universal-router';
import Router from '../Router';
import { renderToStaticMarkup } from 'react-dom/server';
import { RouteParamsInterface } from '../Interfaces';
interface TestContext {
  text?: string;
}
const TestComponent = () => (
  <div>
    <h1>Testing component</h1>
  </div>
);
describe('test router', () => {
  it('test init simple route [/] | [/home/test]', async () => {
    const routes = [
      {
        path: '/',
        action: () => (
          <div>
            <h1>Hello World</h1>
          </div>
        )
      },
      {
        path: '/home/test',
        action: () => (
          <div>
            <h1>Hello World 2!</h1>
          </div>
        )
      }
    ];
    const router = new Router(routes);
    const data1 = await router.resolve({
      pathname: '/'
    });
    const data2 = await router.resolve({
      pathname: '/home/test'
    });
    expect(data1).toMatchSnapshot();
    expect(data2).toMatchSnapshot();
  });

  it('test init route with context params [text: hello world with params!]', async () => {
    const routes = [
      {
        path: '/',
        action: (params: RouteParamsInterface<TestContext>) => (
          <div>
            <h1>{params.context.text}</h1>
          </div>
        )
      }
    ];
    const router = new Router(routes);
    const data = await router.resolve({
      pathname: '/',
      context: {
        text: 'hello world with params!'
      }
    });
    expect(data).toMatchSnapshot();
  });
  it('test add 2 routes [/] | [/another]', async () => {
    const AnotherComponent = () => (
      <h1>Test</h1>
    );
    const router = new Router();    
    router
      .add('/', TestComponent)
      .add('/another', AnotherComponent);
    const data1 = await router.resolve({
      pathname: '/'
    });
    const data2 = await router.resolve({
      pathname: '/another'
    });
    expect(data1).toMatchSnapshot();
    expect(data2).toMatchSnapshot();
  });
  it('test add route with context data [text: test add with data]', async () => {
    const router = new Router();
    router.add('/', TestComponent);

    const data = await router.resolve({
      pathname: '/',
      context: {
        text: 'test add with data'
      }
    });
    expect(data).toMatchSnapshot();
  });
  it('test route with url param /:user', async () => {
    const router = new Router();
    router.add('/:user', TestComponent);

    const data = await router.resolve({
      pathname: '/testing'
    });

    expect(data).toMatchSnapshot();
  });
  it('test state component', async () => {
    class TestStateCompoent extends React.Component {
      constructor(props: RouteParamsInterface<TestContext>) {
        super(props);
        this.state = {
          acitve: false
        };
      }
    }
    const router = new Router();
    router.add('/', TestStateCompoent);

    const data = await router.resolve<TestContext>({
      pathname: '/',
      context: {
        text: 'hello world componet class!'
      }
    });

    expect(data).toMatchSnapshot();
  });
});