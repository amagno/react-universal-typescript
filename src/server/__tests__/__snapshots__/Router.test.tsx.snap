// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test router test add 2 routes [/] | [/another] 1`] = `
<TestComponent
  context={Object {}}
  params={Object {}}
  pathname="/"
/>
`;

exports[`test router test add 2 routes [/] | [/another] 2`] = `
<AnotherComponent
  context={Object {}}
  params={Object {}}
  pathname="/another"
/>
`;

exports[`test router test add route with context data [text: test add with data] 1`] = `
<TestComponent
  context={
    Object {
      "text": "test add with data",
    }
  }
  params={Object {}}
  pathname="/"
/>
`;

exports[`test router test init route with context params [text: hello world with params!] 1`] = `
<div>
  <h1>
    hello world with params!
  </h1>
</div>
`;

exports[`test router test init simple route [/] | [/home/test] 1`] = `
<div>
  <h1>
    Hello World
  </h1>
</div>
`;

exports[`test router test init simple route [/] | [/home/test] 2`] = `
<div>
  <h1>
    Hello World 2!
  </h1>
</div>
`;

exports[`test router test route with url param /:user 1`] = `
<TestComponent
  context={Object {}}
  params={
    Object {
      "user": "testing",
    }
  }
  pathname="/testing"
/>
`;

exports[`test router test state component 1`] = `
<TestStateCompoent
  context={
    Object {
      "text": "hello world componet class!",
    }
  }
  params={Object {}}
  pathname="/"
/>
`;
